types/phoenix_live_view/socket_options.d.ts

```diff
@@ -0,0 +1,206 @@
+   *
+   * Defaults to WebSocket with automatic LongPoll fallback if WebSocket is not defined.
+   * To fallback to LongPoll when WebSocket attempts fail, use `longPollFallbackMs: 2500`.
+   */
+  transport: any; // DT - this should be a type that has a constructor with a URL argument.
+  /*
+   * @param {Function} [opts.longPollFallbackMs] - The millisecond time to attempt the primary transport
+   * before falling back to the LongPoll transport. Disabled by default.
+   */
+  longPollFallbackMs: number;
+  /*
+   * @param {Function} [opts.debug] - When true, enables debug logging. Default false.
+   */
+  debug: boolean;
+  /*
+   * @param {Function} [opts.encode] - The function to encode outgoing messages.
+   * Defaults to JSON encoder.
+   */
+  encode: (msg: object, callback: (encoded: string) => any) => any;
+  /*
+   * @param {Function} [opts.decode] - The function to decode incoming messages.
+   * Defaults to JSON:
+   *
+   * ```javascript
+   * (payload, callback) => callback(JSON.parse(payload))
+   * ```
+   */
+  decode: (payload: any, callback: (decoded: any) => any) => any;
+  /*
+   * @param {number} [opts.timeout] - The default timeout in milliseconds to trigger push timeouts.
+   * Defaults to `DEFAULT_TIMEOUT`
+   */
+  timeout: number;
+  /*
+   * @param {number} [opts.heartbeatIntervalMs] - The millisec interval to send a heartbeat message
+   */
+  heartbeatIntervalMs: number;
+  /*
+   * @param {number} [opts.reconnectAfterMs] - The optional function that returns the millisec
+   * socket reconnect interval. Defaults to stepped backoff of:
+   *
+   * ```javascript
+   * function(tries){
+   *   return [10, 50, 100, 150, 200, 250, 500, 1000, 2000][tries - 1] || 5000
+   * }
+   * ````
+   */
+  reconnectAfterMs: (tries: number) => number;
+  /*
+   * @param {number} [opts.rejoinAfterMs] - The optional function that returns the millisec
+   * rejoin interval for individual channels.
+   *
+   * ```javascript
+   * function(tries){
+   *   return [1000, 2000, 5000][tries - 1] || 10000
+   * }
+   * ````
+   */
+  rejoinAfterMs: (tries: number) => number;
+  /*
+   * @param {Function} [opts.logger] - The optional function for specialized logging, ie:
+   *
+   * ```javascript
+   * function(kind, msg, data) {
+   *   console.log(`${kind}: ${msg}`, data)
+   * }
+   * ```
+   */
+  logger: (kind: string, msg: any, data: any) => void;
+  /*
+   * @param {number} [opts.longpollerTimeout] - The maximum timeout of a long poll AJAX request.
+   * Defaults to 20s (double the server long poll timer).
+   */
+  longpollerTimeout: number;
+  /*
+   * @param {string} [opts.binaryType] - The binary type to use for binary WebSocket frames.
+   * Defaults to "arraybuffer"
+   */
+  binaryType: string;
+  /*
+   * @param {vsn} [opts.vsn] - The serializer's protocol version to send on connect.
+   * Defaults to DEFAULT_VSN.
+   */
+  vsn: string;
```

Comment:
It strikes me as more maintainable to take these from `SocketConnectOption` in the Phoenix types, but I agree they are correct.

Reply:
I will try a version with `extends Phoenix.SocketConnectionOption` locally and see what happens.

Reply:
Oops. I tried this:

```
export default interface SocketOptions 
  extends Partial<SocketConnectOption>, Partial<LiveSocketOption> {
}
```

But it complains because `SocketConnectOption` types `params` as `object | (() => object)`, whereas for the LiveSocket `params` is supposed to be `object | ((el: HTMLElement) => object)`. So unless you know a way to resolve this, I will leave it as is.

Reply:
I guess I could try this:

```
export default interface SocketOptions 
  extends Partial<Omit<SocketConnectOption, "params">>, Partial<LiveSocketOption> {
}
```

Reply:
Nope. That resolves the `params` issue but now I get:

```
'longPollFallbackMs' does not exist in type 'SocketOptions'
```

Doesn't seem to be picking up the `extends Partial<Omit<SocketConnectOption, "params">>` part... I am going to leave it alone; maybe we can revisit in the future.

Reply:
Fair enough!